### PR TITLE
fix: make named axis thread-safe and start testing using multiple threads in CI

### LINKED ISSUE #3802: Named axis is not thread-safe
### Version of Awkward Array

HEAD

### Description and code to reproduce

Named axis uses this class-level dictionary:
https://github.com/scikit-hep/awkward/blob/a7f7d447a64f7e19aff0f488ebf6b4e9220ca850/src/awkward/_namedaxis.py#L32
which is not thread-safe

### LINKED ISSUE #3803: Start testing Awkward Array with parallel threads in CI
Use [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel) to run the awkward tests with parallel threads in python 3.14t

### PR DESCRIPTION
**Pull request by ikrommyd**
Fixes https://github.com/scikit-hep/awkward/issues/3802
Closes https://github.com/scikit-hep/awkward/issues/3803

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -54,7 +54,7 @@ jobs:
           - python-version: '3.14'
             python-architecture: x64
             runs-on: macos-15-intel
-            dependencies-kind: nogil
+            dependencies-kind: basic
```
Comment (ikrommyd) [inline .github/workflows/test.yml:57]: I guess we were using nogil here for a "basic" set of requirements but I want to use jax in nogil now which doesn't have builds for x86 macos. I created a new "basic" requirements for this.

File: requirements-test-nogil.txt [hunk 2]
```diff
@@ -1,5 +1,8 @@
 fsspec>=2022.11.0;sys_platform != "win32"
+jax[cpu]>=0.2.15;sys_platform != "win32"
+numexpr>=2.7
 pandas>=0.24.0;sys_platform != "win32"
+pyarrow>=12.0.0;sys_platform != "win32"
```
Comment (ikrommyd) [inline requirements-test-nogil.txt:5]: Can't yet test numba and safetensors in nogil. Those are the only ones remaining to test in nogil vs full

File: tests/test_2433_copy_behaviors.py [hunk 3]
```diff
@@ -9,6 +9,8 @@
 
 
 def test():
+    behavior = {}
```
Comment (ikrommyd) [inline tests/test_2433_copy_behaviors.py:12]: had to use a local dict within the test otherwise all the threads try to modify the global `ak.behavior` dict.

File: tests/test_2596_named_axis.py [hunk 4]
```diff
@@ -1084,6 +1084,7 @@ def test_negative_named_axis_ak_cartesian():
     ).named_axis == {"x": -2, "y": -1}
 
 
+@pytest.mark.filterwarnings("ignore:The global interpreter lock")
```
Comment (ikrommyd) [inline tests/test_2596_named_axis.py:1087]: `pandas` raises this warning which is treated as error due to global pytest settings
```
>>> import pandas
<frozen importlib._bootstrap>:491: RuntimeWarning: The global interpreter lock (GIL) has been enabled to load module 'pandas._libs.pandas_parser', which has not declared that it can run safely without the GIL. To override this behavior and keep the GIL disabled (at your own risk), run with PYTHON_GIL=0 or -Xgil=0.
```
so we ignore it here. I think it will be fixed in the nest pandas release and this module will be thread-safe too.

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -79,6 +79,10 @@ jobs:
             python-architecture: x64
             runs-on: ubuntu-latest
             dependencies-kind: nogil
+          - python-version: '3.14t'
+            python-architecture: arm64
+            runs-on: macos-14
+            dependencies-kind: nogil
```
Comment (ikrommyd) [inline .github/workflows/test.yml:85]: We start testing arm64 macos in 3.14t too as that is a different architecture and a very common one.

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -165,6 +169,12 @@ jobs:
           python -m pytest -vv -rs tests
         if: startsWith(matrix.python-version, 'pypy')
 
+      - name: Test non-kernels (Python) in multiple threads
+        if: matrix.python-version == '3.14t'
+        run: |
+          python -m pip install 'pytest-run-parallel>=0.8.2'
+          python -m pytest -vv -rs tests --parallel-threads 4 --ignore=tests/test_3364_virtualarray.py --ignore=tests/test_3451_virtualarray_with_jax.py --ignore=tests/test_3475_virtualarray_unknown_length.py
+
```
Comment (ikrommyd) [inline .github/workflows/test.yml:177]: Have to ignore these virtual array tests for now cause they all use fixtures which are shared among all threads in `pytest-run-parallel`. I don't want to use this PR to modify 15K lines of tests to create arrays inside the test functions instead of using fixtures. This will also probably double the lines of each file 🤣

File: tests/test_2433_copy_behaviors.py (hunk 3 above)
Comment (ianna) [reply tests/test_2433_copy_behaviors.py:12]: I think that is intentional - hence the name of the test

Comment (ikrommyd) [reply tests/test_2433_copy_behaviors.py:12]: `ak.behavior` is just a dict: https://github.com/scikit-hep/awkward/blob/a7f7d447a64f7e19aff0f488ebf6b4e9220ca850/src/awkward/__init__.py#L52
So the test tests the same thing with this versus using the awkward defined `ak.behavior` dict.
We do this (keep behaviors in separate outside dicts) all the time in coffea for example. The integrity of the test does not change here.
The problem here is that the test adds behaviors to a dict and checks that addition raises an error before adding the behavior for addition. However, if another thread has added the behavior first in the global dict, then this doesn't raise an error. That's why I wanted the behavior dict be defined within the test.
The other solution would be to just skip the test with multiple threads.

File: src/awkward/_namedaxis.py [hunk 7]
```diff
@@ -24,12 +25,13 @@
 )
 
 
-# just a class for inplace mutation
-class NamedAxis:
-    mapping: AxisMapping
+class _NamedAxisLocal(threading.local):
+    def __init__(self):
+        super().__init__()
+        self.mapping = {}
```
Comment (ianna) [inline src/awkward/_namedaxis.py:30]: This approach is safe for Python threads, but its behavior changes once native threads are involved, which makes it harder to reason about in mixed parallel contexts. Compared to a class-level attribute, `threading.local()` adds extra indirection on every access. That’s usually fine, but it can become noticeable in performance-sensitive code.

Comment (ikrommyd) [reply src/awkward/_namedaxis.py:30]: My understanding, even though @pfackeldey implemented named axis, is that this is used to hold the named axis mapping of the array when you do a `__getitem__` on a layout just to re-add the mapping on the output layout once the `__getitem__` finishes. It's not performance critical.

Comment (pfackeldey) [reply src/awkward/_namedaxis.py:30]: > This approach is safe for Python threads, but its behavior changes once native threads are involved, which makes it harder to reason about in mixed parallel contexts. Compared to a class-level attribute, `threading.local()` adds extra indirection on every access. That’s usually fine, but it can become noticeable in performance-sensitive code.

Python threads are OS threads afaik (I'm not sure what can be mixed here with "native" threads? I may not understand that correctly...). The implementation of @ikrommyd is correct (in my view) for making mutable objects threading-friendly, and is something you can find throughout many large scale OSS (e.g.: [multi-threading-friendly tracing in JAX](https://github.com/jax-ml/jax/blob/643018068a0152500b0e9e8e6f08c0b11f85a9ae/jax/_src/core.py#L1298)). It's also not performance sensitive code here for us (and apart from that I doubt it's much overhead anyways).

Comment (ianna) [reply src/awkward/_namedaxis.py:30]: > > This approach is safe for Python threads, but its behavior changes once native threads are involved, which makes it harder to reason about in mixed parallel contexts. Compared to a class-level attribute, `threading.local()` adds extra indirection on every access. That’s usually fine, but it can become noticeable in performance-sensitive code.
> 
> Python threads are OS threads afaik (I'm not sure what can be mixed here with "native" threads? I may not understand that correctly...). 

You are correct: Python threads (from the `threading` module) are OS threads. However, there is a massive difference between a thread managed by the Python Interpreter and a "naked" thread (which has no "Python identity") managed by a native library (like `OpenMP` or `TBB`).

> The implementation of @ikrommyd is correct (in my view) for making mutable objects threading-friendly, and is something you can find throughout many large scale OSS (e.g.: [multi-threading-friendly tracing in JAX](https://github.com/jax-ml/jax/blob/643018068a0152500b0e9e8e6f08c0b11f85a9ae/jax/_src/core.py#L1298)). It's also not performance sensitive code here for us (and apart from that I doubt it's much overhead anyways).

From what I see JAX uses `threading.local` for short-lived control context during tracing, and then copies that state into explicit configuration before entering native execution. This works because the thread-local state is never accessed once the GIL is released I think. Am I wrong?

There the thread-local object appears to act as shared mutable state rather than transient control context, which means the same guarantees don’t apply if this code ever runs alongside native parallelism (e.g. `OpenMP`, `TBB`, `Numba`, `RDataFrame`, or `cppyy`, and possibly `NumPy` in some cases and definitely in their future releases with `OpenMP`).

Unlike JAX that follows a strict functional/transformational model, by using `Numba`, `RDataFrame`, or `cppyy` we are moving into a stateful/imperative world where the "Magic of the GIL" no longer protects us. That is why for me personally fail-loud beats silent corruption.

Comment (ikrommyd) [reply src/awkward/_namedaxis.py:30]: This named axis dict is never modified by any other compiled library. It is only modified by python. It's never passed into any external libraries or kernels.

Comment (pfackeldey) [reply src/awkward/_namedaxis.py:30]: I second @ikrommyd comment. There's no silent corruption here... This is just a pure python object that is in-place mutated (which is why we need to make it thread local for python as @ikrommyd did); it's only seen by python and not by any other compiled program. That's all there is to it.

Comment (ianna) [reply src/awkward/_namedaxis.py:30]: Thanks for clarification. Shall I assume the users do not (and will not) use the named axis in Numba?

Something like:

```python
from numba import njit, prange, typed

axis_map = typed.Dict.empty(
   key_type=numba.types.unicode_type,
   value_type=numba.types.int64
)
axis_map["x"] = 0
axis_map["y"] = 1

@njit(parallel=True)
def compute_coords_numba(data, axis_map):
    n = data.shape[0]
    result = np.zeros(n)
    for i in prange(n):
        idx_x = axis_map["x"]
        result[i] = data[i, idx_x] * 2.0
    return result
   
import numpy as np
data = np.random.rand(1000, 2)
out = compute_coords_numba(data, axis_map)
print(out[:10])
```

Comment (pfackeldey) [reply src/awkward/_namedaxis.py:30]: The named axis information lives in `.attrs`, so it is as much supported as `.attrs` is. _This_ `NamedAxis` python object here is also only an internal helper and not how one interacts with the named axis interface (see [more details in the docs for how to use them with high-level awkward operations](https://awkward-array.org/doc/main/user-guide/how-to-array-properties-named-axis.html)). If one wants to manually use any metadata (e.g. via `numba.typed.Dict`) in numba then it is numba's `typed.Dict` responsibility to be thread-safe, not awkward-array's.

Comment (ikrommyd) [reply src/awkward/_namedaxis.py:30]: Additionally, all high level functions are not recognized by numba and therefore can't exist in numba compiled functions. So the public named axis API with attrs and `ak.with/without_named_axis` cannot ever exist in numba functions.

**Review by pfackeldey — APPROVED** [review]
This looks good to me know! I'm approving the python side here that makes things thread-safe. @ariostas could you review/approve the CI part of this PR?

**Review by ariostas — APPROVED** [review]
The CI side looks good to me! It does make sense to separate "nogil" and "basic". We should have done that before. And yeah, we should have been testing 3.14t on macOS.

File: src/awkward/_namedaxis.py (hunk 7 above)
Comment (ianna) [reply src/awkward/_namedaxis.py:30]: Thanks for your patience! Please, check if it's clearly documented. Thanks

Comment (ikrommyd) [reply src/awkward/_namedaxis.py:30]: I think our numba documentation is pretty clear that one can't use highlevel features here: https://awkward-array.org/doc/main/user-guide/how-to-use-in-numba-features.html#cannot-use-ak-functions-or-ufuncs
It also just does not compile.

Comment (ianna) [reply src/awkward/_namedaxis.py:30]: Thanks for checking! ✅

**Comment by ikrommyd** [comment]
@henryiii said on slack this is good. Can we merge it?

**Comment by ianna** [comment]
> @henryiii said on slack this is good. Can we merge it?

Sure, go ahead 🙌
